Lookup PX api port from k8s service

harsh-px opened the pull request:
Fixes #70033

Signed-off-by: Harsh Desai <harsh@portworx.com>

**What type of PR is this?**
> /kind bug

**What this PR does / why we need it**: Refer to https://github.com/kubernetes/kubernetes/issues/70033

**Which issue(s) this PR fixes** :
Fixes #70033

**Special notes for your reviewer**:

```release-note
Fixes an issue where Portworx volumes cannot be mounted if 9001 port is already in use on the host and users remap 9001 to another port.
```

Commit:
Lookup PX api port from k8s service

Fixes #70033

Signed-off-by: Harsh Desai <harsh@portworx.com>

idealhack commented:
/kind bug
/ok-to-test

jsafrane commented on pkg/volume/portworx/portworx_util.go:
```diff
@@ -330,3 +335,14 @@ func (util *PortworxVolumeUtil) getPortworxDriver(volumeHost volume.VolumeHost,
+func lookupPXAPIPortFromService(svc *v1.Service) int32 {
+	for _, p := range svc.Spec.Ports {
+		if p.Port == osdMgmtDefaultPort {
+			return p.TargetPort.IntVal
```
What if the service uses named port?

harsh-px replied:
Portworx users use a fixed portworx-service spec generated from https://install.portworx.com. Hence above logic that looks for `port` and `targetPort` works.

The spec is 

```
kind: Service
apiVersion: v1
metadata:
  name: portworx-service
  namespace: kube-system
  labels:
    name: portworx
spec:
  selector:
    name: portworx
  ports:
    - name: px-api
      protocol: TCP
      port: 9001
      targetPort: 9001
```

jsafrane commented on pkg/volume/portworx/portworx_util.go:
```diff
@@ -280,8 +280,31 @@ func createDriverClient(hostname string) (*osdclient.Client, error) {
+	}
+
+	opts := metav1.GetOptions{}
+	svc, err := kubeClient.CoreV1().Services(api.NamespaceSystem).Get(pxServiceName, opts)
```
This issues API call for **each** portworx mount / unmount in kubelet. Can the port number change in a running cluster? Can be the port cached in any way?

harsh-px replied:
@jsafrane I've improved this in the incremental commit.

Now I use an existing saved client as long as it's valid. If port changes, a new client will be created and used subsequently.

I have also refactored the methods and created separate methods for getting a client from local calls (mount, unmount, attach and detach) vs cluster wide calls.

Commit:
cache portworx API port

- reused client whenever possible
- refactor get client function into explicit cluster-wide and local functions

Signed-off-by: Harsh Desai <harsh@portworx.com>

harsh-px commented:
/retest

harsh-px commented:
/test pull-kubernetes-e2e-kops-aws

harsh-px commented:
@jsafrane Can you take a second look and see if it addressed your comments. Let me know !

Commit:
Update portworx to move from glog to klog

Signed-off-by: Harsh Desai <harsh@portworx.com>

jsafrane commented:
/lgtm
/approve
